### ISSUE TITLE
systemd-gpt-auto-generator[188]: /dev/sda: failed to determine partition table type

### ISSUE THREAD

**Issue by ghost** [systemd/systemd#2483]
Running Debian testing (fully updated) using standard repository's systemd version 228-4+b1.

Reported from previous version of systemd in 765586@bugs.debian.org.

It was suggested that -- since bug is still reproducible --  I report this upstream.

Minor bug with no obvious (to me) consequences. Every time system is booted lines such as the following appear on-screen and in dmesg:

[    4.902248] systemd-gpt-auto-generator[188]: /dev/sda: failed to determine partition table type: Input/output error
[    5.562027] systemd[186]: /lib/systemd/system-generators/systemd-gpt-auto-generator failed with error code 1.

I am running four Debian testing systems partitioned and configured almost identically, software-wise. Only one system, a Lenovo Thinkpad 520i, exhibits this problem. Other systems are all older and not using SATA drives.

I suppose the controller / drive are implicated in some way. From lspci command:

00:1f.2 SATA controller: Intel Corporation 6 Series/C200 Series Chipset Family 6 port SATA AHCI Controller (rev 04)

Please let me know if I can provide information which might be useful

**Comment by mbiebl** [comment]
The followup reply at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=765586#10 has more details. Quoting the full reply:

I gathered following possibly helpful information, attached here:

On Wed, 15 Oct 2014, Sven Joachim wrote:

> I don't think there is actually an I/O error here, looking at the code
> systemd-gpt-auto-generator makes this error up:

```
         errno = 0;
         r = blkid_probe_lookup_value(b, "PTTYPE", &pttype, NULL);
         if (r != 0) {
                 if (errno == 0)
                         errno = EIO;
                 log_error("Failed to determine partition table type of %s: %m", node);
                 return -errno;
```

> Somebody who is familiar with libblkid (i.e. not me) might explain why
> blkid_probe_lookup_value() apparently failed but did not set errno.

Great catch. Yeah, blkid_probe_lookup_value apparently just returns -1
on all errors, regardless of what the error was.

This is probably a bug in systemd-gpt-auto-generator, but upstream (and
the maintainer) would know much more than

**Comment by poettering** [comment]
It would be interesting to know what the blkid tool outputs when run on that disk.

**Comment by ghost** [comment]
On 01/31/2016 08:52 AM, Lennart Poettering wrote:

> It would be interesting to know what the blkid tool outputs when run on that disk.

Huh. I really didn't expect this.

~# blkid /dev/sda
/dev/sda: TYPE="zfs_member"

~# blkid /dev/sda1
/dev/sda1: UUID="8e9da3fd-7fc5-47ab-b6a1-bab3f1efc213" TYPE="ext4" 
PARTUUID="fd52ad3f-01"

~# lsblk /dev/sda
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0 465.8G  0 disk
├─sda1   8:1    0   450G  0 part /
├─sda2   8:2    0     1K  0 part
└─sda5   8:5    0  15.8G  0 part [SWAP]

Set up partitioning with default from debian-install rather than 
manually partitioning. Don't know whether or not this is typical for 
SATA drives, since this is the only one I have.

**Comment by poettering** [comment]
Sounds really like something to fix in libblkid then. It appears to have misidentified something here. @karelzak can you look into this, please?

Closing, since it appears to be no issue in systemd itself.

**Comment by mbiebl** [comment]
@poettering shouldn't systemd-gpt-auto-generator at least produce a better error message? input/output error is very confusing in such a case.

**Comment by karelzak** [comment]
ZFS detection is so complex and it requires ZFS magic strings on 4 different offsets  that I have doubts that the disk have never been formatted as ZFS.

I guess someone has used it for ZFS and them created partition table on the device.

See for more details:
 # blkid -p -o udev /dev/sda

erase unwanted (see wipefs man page before you wipe):
  # wipefs --backup --types zfs_member /dev/sda

**Comment by ghost** [comment]
On 02/01/2016 07:57 AM, Karel Zak wrote:

> ZFS detection is so complex and it requires ZFS magic strings on 4
> different offsets  that I have doubts that the disk have never been
> formatted as ZFS.
> 
> I guess someone has used it for ZFS and them created partition table
> on the device.

If I did so, it wasn't on purpose! AFAIR I just used the standard 
partitioning features of debian-installer in a netinst image to 
partition the disk.

> See for more details: # blkid -p -o udev /dev/sda
> 
> erase unwanted (see wipefs man page before you wipe): # wipefs
> --backup --types zfs_member /dev/sda

This is interesting. I appreciate the information. The error message is 
only a tiny annoyance at boot time, but it will be nice to be rid of it!

Best regards,
JP

**Comment by poettering** [comment]
@mbiebl true, we should generate a better error message...

**Comment by poettering** [comment]
@karelzak what precisely does blkid_probe_lookup_value() returning != mean? How would I distuingish an I/O error occuring during this call from the case "this field does not exist"? Or can this function never fail?

@karelzak hmm, isn't there a bug in blkid after all? As I understand there's an MBR partition table on the same disk. Shouldn't blkid have returned with -2, indicating an ambigius result, as both ZFS and MBR should have been detected?

**Comment by mbiebl** [comment]
@poettering thanks for reopening it. it's the little details that show whether we care for our users or not.

**Comment by karelzak** [comment]
@poettering blkid_probe_lookup_value() != 0 means: not found the value in the probing result, nothing else. 

blkid_do_safeprobe() returns -2 for ambivalent result. Unfortunately, currently the library does not care about PT and FS collisions. It cares about conflicts between filesystems only.

It's unusual but in crazy MBR world it's possible to have PT and filesystem on the first device, for example some boot media still uses this layout.

Anyway, you can detect this situation, in the probing result will be "TYPE" (=filesystem) as well as "PTTYPE" (=partition table) value.

Maybe we can add some extra flag to check this problem within library, I'll think about it. (Users are sensitive to these changes, because their messy systems which they use for years stop work after update:-)

**Comment by karelzak** [comment]
@poettering I/O error is -1 returned by blkid_do_safeprobe(), and proper errno.

**Comment by karelzak** [comment]
Note that core of the problem is that partitioning tools don't  wipe old signatures (or at least warn users about possible collisions). The current fdisk warn about it, the next version will also wipe the device. It's also system installer responsibility to prepare (wipe) the devices before it creates a new stuff. For example Fedora/RHEL installer cares about it.

**Comment by poettering** [comment]
@karelzak btw, I think it would be a good idea to introduce an anonymous enum for the results of blkid_do_safeprobe() which map to 0, 1, -2. That's why C knows the concept after all. Dealing with literal numbers is just weird... i.e. something like:

``` C
enum {
        BLKID_PROBE_NOT_FOUND = 1,
        BLKID_PROBE_AMBIGUOUS = -2,
        BLKID_PROBE_FOUND = 0,
};
```

You get the idea...

I'd also prefer if the library would set errno on all errors, in all functions. i.e. blkid_probe_lookup_value() should really set errno to ENXIO or something like that when it can't find something... That way the API would be much more uniform...

**Comment by poettering** [comment]
There's now a fix waiting in PR #2506. Specifically the new code that looks for PTTYPE looks like this:

https://github.com/poettering/systemd/blob/cb9712492f94153b7ce6fc03d6dd3fd95c87baa5/src/gpt-auto-generator/gpt-auto-generator.c#L643

That means not finding a PTTYPE should be handled cleanly.

**Comment by karelzak** [comment]
OK, added to my TODO list.
